### PR TITLE
[ConnectorsForge] Expanding user authentication possibilities

### PR DESCRIPTION
**Pull request by marek-zuwala**

### CODE REVIEW

File: .bob/rules-connectors-forge-rest/6_deployment_execution.xml [hunk 1]
```diff
@@ -942,7 +942,7 @@ Optional configuration (will use defaults if not specified):
               Runs the OS-specific deployment script with user-provided parameters.
               The script handles:
               - Connecting to remote server via SSH (using your SSH keys - no password prompts)
-              - Creating container from pre-built Docker image (ghcr.io/marek-zuwala/connectors-forge:1.0.5.3)
+              - Creating container from pre-built Docker image (ghcr.io/marek-zuwala/connectors-forge:1.0.6.4)
```
Comment (piotr-rachwalik) [inline .bob/rules-connectors-forge-rest/6_deployment_execution.xml]: Should these images be stored in some common registry not your private one?

File: sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/JsonToRecordStream.java [hunk 2]
```diff
@@ -80,76 +81,35 @@ public class JsonToRecordStream implements Iterator<Record>, Closeable
...
-     *            optional authentication headers to include in the HTTP request (may be null)
+     *            optional authentication headers (may be null)
      * @param paginationConfig
      *            optional pagination configuration (may be null for non-paginated APIs)
+     * @param acceptHeader
+     *            value for the HTTP {@code Accept} header (e.g. {@code "application/json"})
      */
     public JsonToRecordStream(String url, String dataPath, List<RestFieldDefinition> fieldDefs,
-            java.util.Map<String, String> authHeaders, PaginationConfig paginationConfig)
+            java.util.Map<String, String> authHeaders, PaginationConfig paginationConfig,
```
Comment (piotr-rachwalik) [inline sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/JsonToRecordStream.java]: Please use imports instead of direct declarations.

File: sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/RestApiMappingLoader.java [hunk 3]
```diff
@@ -178,14 +182,108 @@ private static String parseBaseUrl(JsonNode root) throws IOException
...
+        case API_KEY:
+            return new AuthConfig(type, java.util.Arrays.asList(
+                    new HeaderDef("api_key", "API Key", "Your API key",
+                            true, "Authorization", "ApiKey $api_key")));
+        case OAUTH2:
+            return new AuthConfig(type, java.util.Arrays.asList(
+                    new HeaderDef("bearer_token", "Bearer Token", "OAuth 2.0 access token",
+                            true, "Authorization", "Bearer $bearer_token")));
+        case BASIC:
+            return new AuthConfig(type, java.util.Arrays.asList(
```
Comment (piotr-rachwalik) [inline sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/RestApiMappingLoader.java]: Please use imports instead of direct declarations.

File: sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/RestApiMapping.java [hunk 4]
```diff
@@ -34,23 +35,27 @@ public class RestApiMapping
...
+            String baseUrl, AuthConfig authConfig, String acceptHeader,
+            Map<String, RestTableDefinition> tables, Map<String, String> origin)
     {
         this.connectorName = connectorName;
         this.connectorLabel = connectorLabel;
         this.connectorDescription = connectorDescription;
         this.baseUrl = baseUrl;
-        this.authenticationType = authenticationType != null ? authenticationType : AuthenticationType.NONE;
+        this.authConfig = authConfig != null ? authConfig : new AuthConfig();
+        this.acceptHeader = acceptHeader;
```
Comment (piotr-rachwalik) [inline sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/RestApiMapping.java]: `acceptHeader` is stored as-is with no null/blank guard. `HttpRequest.Builder.header("Accept", value)` throws `NullPointerException` when value is `null` and `IllegalArgumentException` when it is blank — both are unchecked and will surface only at request time.

Please apply the default:

```java
this.acceptHeader = (acceptHeader != null && !acceptHeader.isBlank())
        ? acceptHeader : "application/json";
```

This is consistent with how `authenticationType` was guarded in the old code (`authenticationType != null ? authenticationType : AuthenticationType.NONE`).

File: sdk-gen/subprojects/java/connectors_forge_rest/impl/src/test/java/com/ibm/connect/restconnector/TestAdvancedFeatures.java
Comment (piotr-rachwalik) [inline sdk-gen/subprojects/java/connectors_forge_rest/impl/src/test/java/com/ibm/connect/restconnector/TestAdvancedFeatures.java]: All four existing authentication tests were migrated to the new object form, so the legacy string-form branch in `buildDefaultAuthConfig()` (invoked when a config contains `"$authentication": "api_key"`) now has zero test coverage. This is the primary backward-compatibility guarantee for existing user configs.

Please add at least one test per legacy type, e.g.:

```java
@Test
public void testLegacyStringFormApiKey() throws Exception {
    final String json = "{ \"$hostname\": \"https://x.com\","
            + " \"$authentication\": \"api_key\","
            + " \"$tables\": {} }";
    final RestApiMapping m = RestApiMappingLoader.parse(json);
    assertEquals(AuthenticationType.API_KEY, m.getAuthConfig().getType());
    assertEquals("Authorization",
            m.getAuthConfig().getHeaders().get(0).getHeader());
    assertEquals("ApiKey $api_key",
            m.getAuthConfig().getHeaders().get(0).getValue());
}
```

File: sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/RestInputInteraction.java [hunk 5]
```diff
@@ -180,38 +216,63 @@ private Map<String, String> buildAuthHeaders()
...
+                continue;
             }
-        } else if (authType == AuthenticationType.OAUTH2) {
-            final Object tokenObj = connectionProperties.get("bearer_token");
-            if (tokenObj != null) {
-                headers.put("Authorization", "Bearer " + tokenObj.toString());
-                LOGGER.debug("Using OAuth 2.0 Bearer Token authentication");
-            } else {
-                LOGGER.warn("Bearer token not provided in connection properties");
+            final String resolved = resolveTemplate(hd.getValue(), connectionProperties);
```
Comment (piotr-rachwalik) [inline sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/RestInputInteraction.java]: The new `resolveTemplate()` method handles `$name` substitution and `base64(expr)` encoding — non-trivial logic with several edge cases — but has no unit tests at all. Please add tests covering at least:

- Happy path: single `$var` substitution
- `base64(...)` encoding (e.g. Basic auth round-trip)
- Missing placeholder → returns `null`
- `custom` type end-to-end: multiple headers resolved and placed correctly

File: sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/RestInputInteraction.java [hunk 6]
```diff
@@ -147,28 +154,57 @@ private String buildUrl()
...
-            // Preserve the path prefix from $hostname (e.g. "/api/1.0" in "https://host/api/1.0")
+            // Preserve the path prefix from baseUrl (e.g. "/api/1.0" in "https://host/api/1.0")
             final String basePath = configUrl.getPath();
 
-            // Omit the port when none was written in the config — avoids sending
-            // explicit default ports (e.g. :443) that some servers reject.
-            final String authority = port == -1 ? host : host + ":" + port;
+            // Use host and port from connection properties if supplied; fall back to the config URL.
+            final Object hostProp = connectionProperties.get("host");
+            final Object portProp = connectionProperties.get("port");
```
Comment (piotr-rachwalik) [inline sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/RestInputInteraction.java]: This method now silently overrides `$hostname`'s host and port at runtime using `host`/`port` connection properties. This is a material behaviour change that affects every connector, not just auth-related ones, and is not mentioned anywhere in the PR title or description.

Please either document this change in the PR description (and add a corresponding test), or split it into a separate PR so it can be evaluated in isolation.

File: .bob/rules-connectors-forge-rest/6_deployment_execution.xml (hunk 1 above)
Comment (marek-zuwala) [reply .bob/rules-connectors-forge-rest/6_deployment_execution.xml]: They should and will be as soon as we get access to icr.io. This is a temporary solution while waiting for access

File: sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/RestInputInteraction.java [hunk 7]
```diff
@@ -147,28 +154,57 @@ private String buildUrl()
...
+
+            final String host = (hostProp != null && !hostProp.toString().isBlank())
+                    ? hostProp.toString()
+                    : configUrl.getHost();
```
Comment (piotr-rachwalik) [inline sdk-gen/subprojects/java/connectors_forge_rest/impl/src/main/java/com/ibm/connect/restconnector/RestInputInteraction.java]: Can we have some tests for the host/port override change?

**Review by piotr-rachwalik — COMMENTED** [review]
LGTM

**Review by MichalBodziony — APPROVED** [review]
lgtm
